tools/src/tester/types/eval.types.ts

```diff
@@ -87,16 +87,42 @@ export enum Result {
 export class OutputReference {
   chapter_id: string
   output_name: string
-  private constructor (chapter_id: string, output_name: string) {
+
+  private constructor(chapter_id: string, output_name: string) {
     this.chapter_id = chapter_id
     this.output_name = output_name
   }
 
-  static parse (str: string): OutputReference | undefined {
-    if (str.startsWith('${') && str.endsWith('}')) {
-      const spl = str.slice(2, -1).split('.', 2)
-      return { chapter_id: spl[0], output_name: spl[1] }
+  static parse(str: string): OutputReference[] {
+    const pattern = /\$\{([^}]+)\}/g
+    let match
+    var result = []
+    while ((match = pattern.exec(str)) !== null) {
+      const spl = this.#split(match[1], '.', 2)
+      result.push(new OutputReference(spl[0], spl[1]))
     }
-    return undefined
+    return result
+  }
+
+  static replace(str: string, callback: (chapter_id: any, variable: any) => string): any {
+    // preserve type if 1 value is returned
+    let full_match = str.match(/^\$\{([^}]+)\}$/)
+    if (full_match) {
+      const spl = this.#split(full_match[1], '.', 2)
+      return callback(spl[0], spl[1])
+    } else return str.replace(/\$\{([^}]+)\}/g, (_, k) => {
+      const spl = this.#split(k, '.', 2)
+      return callback(spl[0], spl[1])
+    });
+  }
```

Comment:
Should add a comment explaining what this function does because it's not clear what it's trying to achieve.